fix(deploy): wait if dependency is being deployed

andreafalzetti opened the pull request:
# Proposed changes

Fixes #3281 

Loom video demo: https://www.loom.com/share/8d7c2482a1c54f318b3f712b31c3f0a2

Follow up demo (covers scenario 4): https://www.loom.com/share/1dca9392bce049268ff97b8e44499588

Context: Currently, using `okteto deploy` on a manifest that has dependencies already being deployed, despite having `wait: true` on the dependency, does not wait for the deployment to be completed.

Proposed change: Extending the `deploy` cmd to wait when a repository is found in a "deploying" state. To achieve that, we can rely on an existing Configmap being found and its `Status` value.

Additionally, if we found that a previous job is running, we can also stream the logs. Unfortunately this is only possible for "pipeline deployments". If the deploy occurs locally, the logs won't be available for retrvial.

## How to test

There are 2 scenarios to test.

Scenario 1: **Deploy of the movies-multi-repo twice
Steps:
- Checkout this branch and build the CLI
- Clone movies-multi-repo: `git clone git@github.com:okteto/movies-multi-repo.git`
- Open two terminal sessions and run `okteto deploy` on each one
- Notice how the second one will stream the logs of the first one

Scenario 2: **Deploy `movies-frontend` and `movies-multi-repo`
Steps:
- Checkout this branch and build the CLI
- Clone movies-frontend: `git clone git@github.com:okteto/movies-frontend.git`
- Change the `okteto.yml` of `movies-frontend` removing the `depencencies` block
- Clone movies-multi-repo: `git clone git@github.com:okteto/movies-multi-repo.git`
- Change the `okteto.yml` of `movies-multi-repo` with:
```yaml
dependencies:
  movies-frontend:
    repository: https://github.com/okteto/movies-frontend
    wait: true
deploy:
  - name: Deploy MongoDB
    command: helm upgrade --install mongodb mongodb-12.1.30.tgz -f values.yml
 ```
- Open a terminal sessions for each repo and run `okteto deploy` on each one
- Notice how the second one will say that movies-frontend is already being deployed and waits until it ends.

Scenario 3: **Don't wait for the dependency**
Steps:
- Checkout this branch and build the CLI
- Clone movies-frontend: `git clone git@github.com:okteto/movies-frontend.git`
- Clone movies-multi-repo: `git clone git@github.com:okteto/movies-multi-repo.git`
- Modify the `okteto.yml` manifest of `movies-multi-repo`, set "wait: false`
- Open a terminal sessions for each repo and run `okteto deploy` on each one
- Notice how the second one will say that movies-frontend is already being scheduled for deploy and continue without waiting

Scenario 4: **Recover from `error` state**
Steps:
- Checkout this branch and build the CLI
- Clone movies-frontend: `git clone git@github.com:okteto/movies-frontend.git`
- Clone movies-multi-repo: `git clone git@github.com:okteto/movies-multi-repo.git`
- Modify the `okteto.yml` manifest of `movies-frontend`, adding a non-existing command in the deploy (so that it will break)
- Open a terminal sessions for each repo and run `okteto deploy` on each one
- Notice how the first will go into error state, but when you deploy the second one will recover to progressing and eventually to deployed

andreafalzetti commented on cmd/pipeline/deploy.go:
```diff
@@ -375,7 +421,7 @@ func (o *DeployOptions) setDefaults() error {
 
 func (o *DeployOptions) toPipelineDeployClientOptions() (types.PipelineDeployOptions, error) {
-	varList := []types.Variable{}
+	var varList []types.Variable
```
Change suggested by the linter ✨

jLopezbarb replied:
Which linter are you using? I think we should all start using the same linter to have a common agreement on these things

andreafalzetti replied:
@jLopezbarb it's the one configured in the PR actions, DeepSource: https://github.com/okteto/okteto/pull/3455/checks?check_run_id=12068422538 

Does it block only my PRs? 😂

andreafalzetti replied:
I guess the confusion might be the word "linter" - this seems more like a generic code health solution, but this sort of suggestions generally come from linters - sorry for the confusion :)

andreafalzetti commented on cmd/pipeline/destroy_test.go:
```diff
@@ -25,7 +25,7 @@ import (
 )
 
-func TestDestroyPipelineSuccesful(t *testing.T) {
+func TestDestroyPipelineSuccessful(t *testing.T) {
```
Some typos in this file 🧹

andreafalzetti commented on cmd/pipeline/deploy.go:
```diff
@@ -126,10 +126,56 @@ func (pc *Command) ExecuteDeployPipeline(ctx context.Context, opts *DeployOption
 		if err == nil {
-			oktetoLog.Success("Skipping repository '%s' because it's already deployed", opts.Name)
-			return nil
+			if cfg != nil && cfg.Data != nil {
```
I've haven't added tests of the function `ExecuteDeployPipeline` because it does have  dependencies which make it hard to add them, and it'd require a refactor.

I suggest having a follow-up issue for this PR to schedule for next sprint for example, to refactor this function and make it easier to test

jLopezbarb replied:
Maybe we can encapsulate this new code into a function. Probably that makes this new code testable without having to test `ExecuteDeployPipeline` function

andreafalzetti replied:
ok I've tried the idea of extracting the new code into a function, however it became very complicated too, especially testing all the stuff that is a dependency like `configmaps.WaitForStatus` and `pc.waitUntilRunning` without having to refactor a huge amout.

It would probably be quicker and safer to takle this as an ad-hoc refactor, rewriting some code from scratch..

andreafalzetti commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+	k8sTesting "k8s.io/client-go/testing"
+)
+
+func TestGet(t *testing.T) {
```
This includes tests for the new code and also for the `Get` function. It felt overwelming to add testing for everything but it's something we can follow-up on with a dedicated issue.

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+	clientset := fake.NewSimpleClientset(mockCfg)
+
+	cfg, err := configmaps.Get(ctx, mockCfgName, ns, clientset)
+	if err != nil {
```
NIT: We usually use https://github.com/stretchr/testify to tests the code. There is no need to have lots of if statements and it also provides what's the difference between both objects
```suggestion
	require.NoError(t, err)
```

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+		t.Fatalf("unexpected error returned by Get(): %v", err)
+	}
+
+	if mockCfgName != cfg.Name {
```
```suggestion
	assert.Equal(t, mockCfgName, cfg.Name)
```

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+		t.Errorf("Expected cfg name %s, got %s", mockCfgName, cfg.Name)
+	}
+
+	if mockCfgStatus != cfg.Data["status"] {
```
```suggestion
	assert.Equal(t,mockCfgStatus, cfg.Data["status"] )
```

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+	})
+
+	cfg, err := configmaps.Get(ctx, "fake-cfg-name", "fake-ns", clientset)
+	if cfg != nil {
```
```suggestion
assert.Nil(t, cfg)
```

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+		t.Errorf("Expected cfg to be nil, but got: %v", cfg)
+	}
+
+	if err == nil || err.Error() != mockErr.Error() {
```
```suggestion
	assert.ErrorIs(t, mockErr, err)
```

jLopezbarb commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+
+	clientset := fake.NewSimpleClientset(mockCfg)
+	err := configmaps.WaitForStatus(ctx, mockCfgName, ns, "fake-final-status", 1*time.Microsecond, clientset)
+	if err != oktetoErrors.ErrTimeout {
```
```suggestion
	assert.ErrorIs(t, err, oktetoErrors.ErrTimeout)
```

andreafalzetti commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,146 @@
+		clientset.CoreV1().ConfigMaps(ns).Update(ctx, mockCfg, metav1.UpdateOptions{})
+	}()
+
+	err := configmaps.WaitForStatus(ctx, mockCfgName, ns, "fake-final-status", 5*time.Second, clientset)
```
I will have a look at this PR today, for now if you agree we can merge as is and come back to it later once we agreed upon an approach?

andreafalzetti replied:
@jLopezbarb I made a small change but now the tests run super fast, because I am injecting the ticker, which I think it's a nice thing to do anyway, so from the outside the function, it can be controlled how frequently we check of a status change. Have a look at the diff: https://github.com/okteto/okteto/compare/b6a16882a4bbb4064252df190bfb3502c3739983..ceafa44661010c21ecb57b94142aeb5ca6eab6d2

This also removes the issue of running the tests "slowly"

AdrianPedriza commented on pkg/k8s/configmaps/crud_test.go:
```diff
@@ -0,0 +1,137 @@
+	cfg, err := configmaps.Get(ctx, mockCfgName, ns, clientset)
+
+	require.NoError(t, err)
+	assert.Equal(t, mockCfgName, cfg.Name)
```
I think we can compare the whole object here instead compare specifics labels
```suggestion
	assert.Equal(t, mockCfg, cfg)
```

andreafalzetti replied:
good shout, done!

AdrianPedriza commented on pkg/errors/errors.go:
```diff
@@ -271,3 +274,8 @@ func IsClosedNetwork(err error) bool {
 }
+
+// IsTimeout returns true if err is of the type timeout
+func IsTimeout(err error) bool {
```
do we really need this func? Could we use errors.Is() directly?

andreafalzetti replied:
fair point, it had more logic before, but since I've changed to `errors.Is` I agree with you and it can be removed

AdrianPedriza commented:
I can't reproduce scenario 2. When I deploy `movies-frontend` and then `movies-multi-repo` I got the following for `movies-multi-repo`: 
![Captura de pantalla 2023-03-21 a las 11 19 04](https://user-images.githubusercontent.com/22500940/226577412-279887d3-b4b2-4e8b-8317-c79c6f0b6d52.png)
Do I miss something?

andreafalzetti commented:
> I can't reproduce scenario 2 and 3. When I deploy `movies-frontend` and then `movies-multi-repo` I got the following for `movies-multi-repo`: ![Captura de pantalla 2023-03-21 a las 11 19 04](https://user-images.githubusercontent.com/22500940/226577412-279887d3-b4b2-4e8b-8317-c79c6f0b6d52.png) Do I miss something?

@AdrianPedriza I've realised that `movies-multi-repo` by default does not wait for dependencies to be deployed and I've changed the manifest to wait. I've updated the steps in Scenario 2 to reflect this, sorry!

AdrianPedriza commented:
> > I can't reproduce scenario 2 and 3. When I deploy `movies-frontend` and then `movies-multi-repo` I got the following for `movies-multi-repo`:  Do I miss something?
> 
> @AdrianPedriza I've realised that `movies-multi-repo` by default does not wait for dependencies to be deployed and I've changed the manifest to wait. I've updated the steps in Scenario 2 to reflect this, sorry!
> 
> Update: also, another change that Ive added to the instructions is to update `movies-frontend` Okteto's manifest to get rid of the dependencies (because it depends on `movies-multi-repo` itself, it was giving you that error).
> 
> Maybe it would have been easier next time to setup ad-hoc repos in my personal GitHub account for testing the PR. I wanted to keep it simple and reused the ones we have but with all the changes required I am not sure what was more convinent :)

I already got to test the scenarios! 👍
